sample/sample.go

```diff
@@ -0,0 +1,174 @@
+// constant.
+func Rejection(samples []float64, target dist.LogProber, proposal dist.RandLogProber, c float64, src *rand.Rand) (nProposed int, ok bool) {
+	if c < 1 {
+		panic("rejection: acceptance constant must be greater than 1")
+	}
+	f64 := rand.Float64
+	if src != nil {
+		f64 = src.Float64
+	}
+	var idx int
+	for {
+		nProposed++
+		v := proposal.Rand()
+		qx := proposal.LogProb(v)
+		px := target.LogProb(v)
+		accept := math.Exp(px-qx) / c
+		if accept > 1 {
+			// Invalidate the whole result and return a failure
+			for i := range samples {
+				samples[i] = math.NaN()
+			}
+			return nProposed, false
+		}
+		if accept > f64() {
+			samples[idx] = v
+			idx++
+			if idx == len(samples) {
+				break
+			}
+		}
+	}
+	return nProposed, true
+}
+
+type MHProposal interface {
+	// ConditionalDist returns the probability of the first argument conditioned on
+	// being at the second argument
+	//  p(x|y)
+	ConditionalLogProb(x, y float64) (prob float64)
+
+	// ConditionalRand generates a new random location conditioned being at the
+	// location y
+	ConditionalRand(y float64) (x float64)
+}
+
+// MetropolisHastings generates len(samples) samples using the Metropolis Hastings
+// algorithm (http://en.wikipedia.org/wiki/Metropolis%E2%80%93Hastings_algorithm),
+// with the given target and proposal distributions, starting at the intial location
+// and storing the results in-place into samples. If src != nil, it will be used to generate random
+// numbers, otherwise rand.Float64 will be used.
+//
+// Metropolis-Hastings is a Markov-chain Monte Carlo algorithm that generates
+// samples according to the distribution specified by target by using the Markov
+// chain implicitly defined by the proposal distribution. At each
+// iteration, a proposal point is generated randomly from the current location.
+// This proposal point is accepted with probability
+//  p = min(1, (target(new) * proposal(current|new)) / (target(current) * proposal(new|current)))
+// If the new location is accepted, it is stored into samples and becomes the
+// new current location. If it is rejected, the current location remains and
+// is stored into samples. Thus, a location is stored into samples at every iteration.
+//
+// The samples in Metrpolis Hastings are correlated with one another through the
+// Markov-Chain. As a result, the initial value can have a significant influence
+// on the early samples, and so typically, the first sapmles generated by the chain.
+// are ignored. This is known as "burn-in", and can be accomplished with slicing
+//  nSamples := 1000
+//  burnin := 50
+//  samples := make([]float64, nSamples+burnin)
+//  MetropolisHastings(x, intial, target, proposal, src)
+//  x = x[burnin:]
+// The exact length of the burn-in will depend on the sampling and the target
+// distribution.
+func MetropolisHastings(samples []float64, initial float64, target dist.LogProber, proposal MHProposal, src *rand.Rand) {
```

Comment:
Do you mean like the "trim" in Matlab? I don't think so. First of all, it's controversial. Some say you should, some say you shouldn't. It's more code to implement, it creates a more complicated function signature, and there are more failure rates (if trim < 0). Secondly, I don't think it's a good idea. If you are interested in an expected value (or something), you should keep all of your samples as they are information about the function. Trim is only useful when  trying to de-correlate samples for some other purpose. If that's the case, the user can trim themselves
````
trimmed := make([]float64, len(samples) / trim)
for i := 0; i < len(samples) / trim; i++ {
    trimmed[i] = samples[i*trim]
}
````
If you think it's worthwhile to comment that segment, I can. The controversy over trimming seemed too much for the function comment.